src/Common/Services/AppInstallationHandler.cs

```diff
@@ -80,6 +105,94 @@ private async Task ProcessAppInstallation(CustomerEnvironmentApplicationPackage
             }
         }
 
+        private async Task ProcessEnvironmentDeployment(CustomerEnvironment environment, DeploymentTarget target, DirectoryInfo outputDir)
+        {
+            switch (target)
+            {
+                case DeploymentTarget.dockerswarm:
+                case DeploymentTarget.KubernetesOnPremisesTarget:
+                case DeploymentTarget.OpenShiftOnPremisesTarget:
+                    string outputPath = outputDir != null ? outputDir.FullName : Path.Combine(Directory.GetCurrentDirectory(), "out");
+                    bool success = await _artifactsDownloaderHandler.Handle(environment, outputPath);
+                    if (success)
+                    {
+                        _session.LogInformation($"Customer Environment created at {outputPath}");
+                    }
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        private async Task UpdateQueuePositionAsync(long deployableId, CancellationToken token)
+        {
+            bool presentingLoading = false;
+            int? position = null;
+            int initialTopLine;
+            string msg;
+
+            while (!this._isInstallationFinished && !_hasInstallationStarted && !token.IsCancellationRequested)
```

Comment:
"!this._isInstallationFinished && !_hasInstallationStarted" - do we need to have both in this condition. If the installation didn't start, it didn't finish yet.